### ISSUE TITLE
negative Jacobian

### ISSUE THREAD

**Issue by BalticPinguin** [libMesh/libmesh#2359]
Dear all,
I am generating my mesh with tetgen and from time to time the calculation breaks due to 'negative Jacobian', so I would like to finally get this fixed now, but need help in figuring out what actually happens. (I am not sure this is related to #2274 somehow, but at least this is still persistent.)

I created a (not really) [minimum example](https://gist.github.com/BalticPinguin/5e6f7da0b46957a028dc4372664bdc05.js), where I see the error.
Unfortunately, it seems to critically depend on ``all`` coordinates, so when I tried to shorten the list of points, I could not reproduce this failure.

The error I than get is
````
Elem Information
   id()=1, unique_id()=177682, processor_id()=0
   type()=TET4
   dim()=3
   n_nodes()=4
    0  Node id()=4, processor_id()=0, Point=(x,y,z)=(-0.0350807, 0.00516588, -0.0571602)
    DoFs=
    1  Node id()=5, processor_id()=0, Point=(x,y,z)=(-0.0367887, 0.00678192, -0.0559037)
    DoFs=
    2  Node id()=6, processor_id()=0, Point=(x,y,z)=(-0.0338311, 0.00778059, -0.0576155)
    DoFs=
    3  Node id()=7, processor_id()=0, Point=(x,y,z)=(0.0260922, 0.00804063, 0.0647749)
    DoFs=
   n_sides()=4
    neighbor(0)=864761
    neighbor(1)=519169
    neighbor(2)=337711
    neighbor(3)=702062
   hmin()=0.00266602, hmax()=0.13645
   volume()=-1.72416e-07
   active()=1, ancestor()=0, subactive()=0, has_children()=0
   parent()=nullptr
   level()=0, p_level()=0
   refinement_flag()=DO_NOTHING
   p_refinement_flag()=DO_NOTHING
   infinite()=0
   DoFs=
  -0.001708  0.0012496  0.0611729
  0.00161604  0.00261471  0.00287475
  0.0012565  -0.0004553  0.121935
ERROR: negative Jacobian -1.0345e-06 at point index 0 in element 1
Stack frames: 9
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*)
2: libMesh::FEMap::compute_single_point_map(unsigned int, std::vector<double, std::allocator<double> > const&, libMesh::Elem const*, unsigned int, std::vector<libMesh::Node const*, std::allocator<libMesh::Node const*> > const&, bool)
3: libMesh::FEMap::compute_affine_map(unsigned int, std::vector<double, std::allocator<double> > const&, libMesh::Elem const*)
4: libMesh::FEMap::compute_map(unsigned int, std::vector<double, std::allocator<double> > const&, libMesh::Elem const*, bool)
5: libMesh::FE<3u, (libMesh::FEFamily)0>::reinit(libMesh::Elem const*, std::vector<libMesh::Point, std::allocator<libMesh::Point> > const*, std::vector<double, std::allocator<double> > const*)
6: /scratch/tm162/bin/FreeWilly/MinimumErrorExamplee/FrWll-dbg() [0x423908]
7: __libc_start_main
8: /scratch/tm162/bin/FreeWilly/MinimumErrorExamplee/FrWll-dbg() [0x4235ca]
[0] ../master/src/fe/fe_map.C, line 1153, compiled Dec  4 2019 at 10:16:37
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
````
**Note:** the lines after ``DOSs`` is the output of ``dx_dxi``, ``dx_deta``, ``dx_dzeta`` (second and third line with ``dy_dxi, ... ``and ``dz_dxi,...``, respectively); so plotting this element together with these vectors in point 0 shows that the mapping vectors really span a left-system 
[(see this graph)](https://github.com/libMesh/libmesh/files/3926678/element_map.pdf)
Also it seems as if this occurs only (mainly) to such flat elements as shown in the graph (here the dihedral is restricted to 12 degree, so it should not be too flat).

Any ideas of where this error actually comes from are welcome!
Best regards and many thanks in advance, 
Hubert

**Comment by jwpeterson** [comment]
With a TET4 you can only have a negative Jacobian if the element is completely inverted. So it will "look" fine to the naked eye, but it is numbered incorrectly for libmesh. Maybe there is an issue in translating the node ordering from the Tetgen numbering into the libmesh numbering?

**Comment by BalticPinguin** [comment]
yes, this is what I meant by 'left-system'; 
Ok, so you think the error is in the tetgen-interface somewhere? Than I will have a look there. Thank you very much.

**Comment by jwpeterson** [comment]
> Ok, so you think the error is in the tetgen-interface somewhere?

It's possible but I don't think it's likely, unless you are getting a negative Jacobian for every Tet in the mesh? I would expect all the Tets coming out of Tetgen to be numbered consistently (either all left- or all right-handed) so I would expect our Tetgen interface code to either be right for all elements or wrong for all elements.

In your application, are the nodes or elements moving? If so, then perhaps one vertex of the tet passes through an opposite face due to this motion...

**Comment by BalticPinguin** [comment]
No, I have static elements.
so maybe it is a bug inside tetgen if the node order for elements is not changed?

**Comment by jwpeterson** [comment]
Yes, possibly that or maybe Tetgen doesn't care about the node ordering, so we need to inspect each element that we get from Tetgen and permute the ordering if it has a negative Jacobian.

**Comment by BalticPinguin** [comment]
ok, great.
Than I will have a closer look into the tetgen-code first.
Many thanks for the clarification!

**Comment by BalticPinguin** [comment]
Ok, I think I found the solution now: It is tetgens fault, but not a bug: One just must tell it to check the mesh for consistency by adding the option 'C'.

I think this is a reasonable default option, so I'll PR the change in the default interface.

I cannot check this in detail currently since I changed my code to generate other point-distributions to give to tetgen; but I still think we are save to close the issue.

Many thanks again for the explanations and help to find the right place to search for the problem.
